File: backend/app/services/gemini_service.py
import json
import logging
import asyncio
from datetime import datetime
from typing import AsyncGenerator, Optional, Tuple
import httpx
from app.config import settings
from app.database import SessionLocal
from app.models.system_setting import SystemSetting
from app.services import prompt_service

logger = logging.getLogger(__name__)

_cached_api_key: Optional[str] = None


def _init_cached_api_key() -> None:
    global _cached_api_key
    if _cached_api_key is not None:
        return
    try:
        db = SessionLocal()
        try:
            setting = db.query(SystemSetting).filter(SystemSetting.key == "gemini_api_key").first()
            if setting and setting.value and setting.value.strip():
                _cached_api_key = setting.value.strip()
        finally:
            db.close()
    except Exception:
        pass

    if not _cached_api_key and settings.GEMINI_API_KEY:
        _cached_api_key = settings.GEMINI_API_KEY.strip()


_init_cached_api_key()


def get_active_api_key() -> Optional[str]:
    _init_cached_api_key()
    return _cached_api_key


def get_masked_api_key() -> str:
    key = get_active_api_key()
    if not key:
        return ""
    if len(key) <= 8:
        return "*" * len(key)
    return f"{key[:4]}...{key[-4:]}"


def update_api_key(new_key: str) -> bool:
    global _cached_api_key
    if not new_key or not new_key.strip():
        return False
    clean = new_key.strip()
    _cached_api_key = clean

    try:
        db = SessionLocal()
        try:
            setting = db.query(SystemSetting).filter(SystemSetting.key == "gemini_api_key").first()
            if not setting:
                setting = SystemSetting(key="gemini_api_key", value=clean, updated_at=datetime.utcnow())
                db.add(setting)
            else:
                setting.value = clean
                setting.updated_at = datetime.utcnow()
            db.commit()
            logger.info("Gemini API key saved in system_settings database table.")
            return True
        finally:
            db.close()
    except Exception as ex:
        logger.error(f"Failed to save Gemini API key in database: {ex}")
        return False


def remove_api_key() -> bool:
    global _cached_api_key
    _cached_api_key = None
    try:
        db = SessionLocal()
        try:
            setting = db.query(SystemSetting).filter(SystemSetting.key == "gemini_api_key").first()
            if setting:
                db.delete(setting)
                db.commit()
            return True
        finally:
            db.close()
    except Exception as ex:
        logger.error(f"Failed to remove Gemini API key from database: {ex}")
        return False


async def test_api_key_async(test_key: Optional[str] = None) -> Tuple[bool, str, Optional[str]]:
    key_to_use = test_key.strip() if test_key and test_key.strip() else get_active_api_key()
    if not key_to_use:
        return False, "No API key provided to test. Please enter a Gemini API key.", None

    model = settings.GEMINI_MODEL
    url = f"https://generativelanguage.googleapis.com/v1beta/models/{model}:generateContent?key={key_to_use}"

    request_obj = {
        "contents": [{"parts": [{"text": "Respond with 'OK' if you can read this."}]}],
        "generationConfig": {"maxOutputTokens": 10},
    }

    try:
        async with httpx.AsyncClient(timeout=15.0) as client:
            response = await client.post(url, json=request_obj)

            if response.is_success:
                return True, f"Connection successful! {model} is responsive.", model

            try:
                err_data = response.json()
                msg = err_data.get("error", {}).get("message", response.text)
                return False, f"Gemini API Error ({response.status_code}): {msg}", model
            except Exception:
                return False, f"Gemini API returned status {response.status_code}: {response.reason_phrase}", model
    except Exception as ex:
        return False, f"Connection error: {str(ex)}", model


async def generate_content_async(
    prompt: str,
    max_tokens: int = 4096,
    system_instruction: Optional[str] = None,
    response_mime_type: Optional[str] = None,
    thinking_budget: Optional[int] = None,
) -> str:
    api_key = get_active_api_key()
    if not api_key:
        return (
            "Gemini API key is not configured. Please add your Gemini API key using the 'Gemini API Key' "
            "button in the chat header."
        )

    model = settings.GEMINI_MODEL
    url = f"https://generativelanguage.googleapis.com/v1beta/models/{model}:generateContent?key={api_key}"

    generation_config: dict = {
        "maxOutputTokens": max_tokens,
    }

    if thinking_budget is not None:
        generation_config["thinkingConfig"] = {"thinkingBudget": thinking_budget}

    request_obj = {
        "contents": [{"parts": [{"text": prompt}]}],
        "generationConfig": generation_config,
    }

    if response_mime_type:
        request_obj["generationConfig"]["responseMimeType"] = response_mime_type

    if system_instruction:
        request_obj["systemInstruction"] = {
            "parts": [{"text": system_instruction}]
        }

    for attempt in range(3):
        try:
            async with httpx.AsyncClient(timeout=45.0) as client:
                response = await client.post(url, json=request_obj)

                if response.status_code == 429:
                    return (
                        "The AI service has reached its current Gemini API quota. Please wait for the quota "
                        "to reset or use a Gemini API project with available quota."
                    )

                if not response.is_success:
                    error_body = response.text
                    logger.warning(f"Gemini error (attempt {attempt}): {response.status_code} - {error_body}")
                    if response.status_code >= 500 and attempt < 2:
                        await asyncio.sleep(2**attempt)
                        continue
                    return "The AI service could not process the request. Please check the Gemini API configuration."

                data = response.json()
                candidates = data.get("candidates", [])
                if not candidates:
                    return "I couldn't generate an answer from the available information."

                content = candidates[0].get("content", {})
                parts = content.get("parts", [])

                text_parts = []
                for p in parts:
                    if p.get("thought") is True:
                        continue
                    t = p.get("text")
                    if t:
                        text_parts.append(t)

                if text_parts:
                    return "".join(text_parts).strip()
                elif parts:
                    for p in reversed(parts):
                        t = p.get("text")
                        if t:
                            return t.strip()

                return "I couldn't generate an answer from the available information."

        except Exception as ex:
            logger.error(f"Exception calling Gemini API (attempt {attempt}): {ex}")
            if attempt >= 2:
                return "An unexpected error occurred while generating the AI response."
            await asyncio.sleep(2**attempt)

    return "Sorry, Gemini is currently unavailable."


async def stream_content_async(
    prompt: str,
    max_tokens: int = 4096,
    system_instruction: Optional[str] = None,
    thinking_budget: Optional[int] = 0,
) -> AsyncGenerator[str, None]:
    api_key = get_active_api_key()
    if not api_key:
        yield (
            "Gemini API key is not configured. Please add your Gemini API key using the 'Gemini API Key' "
            "button in the chat header."
        )
        return

    model = settings.GEMINI_MODEL
    url = f"https://generativelanguage.googleapis.com/v1beta/models/{model}:streamGenerateContent?alt=sse&key={api_key}"

    generation_config: dict = {"maxOutputTokens": max_tokens}
    if thinking_budget is not None:
        generation_config["thinkingConfig"] = {"thinkingBudget": thinking_budget}

    request_obj = {
        "contents": [{"parts": [{"text": prompt}]}],
        "generationConfig": generation_config,
    }

    if system_instruction:
        request_obj["systemInstruction"] = {
            "parts": [{"text": system_instruction}]
        }

    try:
        async with httpx.AsyncClient(timeout=60.0) as client:
            async with client.stream("POST", url, json=request_obj) as response:
                if not response.is_success:
                    err = await response.aread()
                    logger.error(f"Gemini stream returned error {response.status_code}: {err.decode('utf-8', errors='ignore')}")
                    yield "Error calling Gemini streaming API."
                    return

                async for line in response.aiter_lines():
                    if not line:
                        continue
                    if line.startswith("data: "):
                        json_str = line[6:].strip()
                        if json_str == "[DONE]":
                            break
                        try:
                            node = json.loads(json_str)
                            candidates = node.get("candidates", [{}])
                            if candidates:
                                parts = candidates[0].get("content", {}).get("parts", [])
                                for p in parts:
                                    if p.get("thought") is True:
                                        continue
                                    chunk_text = p.get("text", "")
                                    if chunk_text:
                                        yield chunk_text
                        except Exception:
                            pass
    except Exception as ex:
        logger.error(f"Stream exception: {ex}")
        yield f"Streaming connection error: {str(ex)}"


async def ask_gemini_async(question: str, context: str) -> str:
    prompt = prompt_service.build_chat_prompt(question, context)
    return await generate_content_async(
        prompt, max_tokens=4096, system_instruction=prompt_service.CHAT_SYSTEM_INSTRUCTION
    )


async def ask_summary_async(context: str) -> str:
    prompt = prompt_service.build_summary_prompt(context)
    return await generate_content_async(prompt, max_tokens=4096)


async def ask_notes_async(context: str) -> str:
    prompt = prompt_service.build_notes_prompt(context)
    return await generate_content_async(prompt, max_tokens=4096)


async def ask_quiz_async(
    context: str, difficulty: str = "Medium", questions: int = 10, skills: Optional[list] = None
) -> str:
    prompt = prompt_service.build_quiz_prompt(context, difficulty, questions, skills)
    return await generate_content_async(prompt, max_tokens=8192, response_mime_type="application/json")


async def ask_course_skills_async(course_title: str, context: str) -> str:
    prompt = prompt_service.build_course_skills_prompt(course_title, context)
    return await generate_content_async(prompt, max_tokens=8192, response_mime_type="application/json")

File: backend/Services/GeminiService.cs
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;
using Microsoft.EntityFrameworkCore;
using VideoIntelligencePlatform.Backend.Data;
using VideoIntelligencePlatform.Backend.Models;

namespace VideoIntelligencePlatform.Backend.Services;

public class GeminiService : IGeminiService
{
    private readonly HttpClient _httpClient;
    private readonly AppDbContext _db;
    private static string? _cachedApiKey;
    private readonly string _model;
    private readonly IPromptService _promptService;
    private readonly ILogger<GeminiService> _logger;

    public GeminiService(HttpClient httpClient, AppDbContext db, IConfiguration configuration, IPromptService promptService, ILogger<GeminiService> logger)
    {
        _httpClient = httpClient;
        _db = db;
        _promptService = promptService;
        _logger = logger;
        _model = configuration["GeminiModel"] 
            ?? Environment.GetEnvironmentVariable("GEMINI_MODEL") 
            ?? "gemini-3.5-flash";

        if (_cachedApiKey == null)
        {
            try
            {
                var setting = _db.SystemSettings.AsNoTracking().FirstOrDefault(s => s.Key == "gemini_api_key");
                if (setting != null && !string.IsNullOrWhiteSpace(setting.Value))
                {
                    _cachedApiKey = setting.Value.Trim();
                }
            }
            catch { }

            if (string.IsNullOrWhiteSpace(_cachedApiKey))
            {
                var envKey = configuration["GeminiApiKey"] ?? Environment.GetEnvironmentVariable("GEMINI_API_KEY");
                if (!string.IsNullOrWhiteSpace(envKey))
                {
                    _cachedApiKey = envKey.Trim();
                }
            }
        }
    }

    public string? GetActiveApiKey() => _cachedApiKey;

    public string GetMaskedApiKey()
    {
        if (string.IsNullOrWhiteSpace(_cachedApiKey)) return string.Empty;
        if (_cachedApiKey.Length <= 8) return new string('*', _cachedApiKey.Length);
        return $"{_cachedApiKey[..4]}...{_cachedApiKey[^4..]}";
    }

    public bool UpdateApiKey(string newKey)
    {
        if (string.IsNullOrWhiteSpace(newKey)) return false;
        var clean = newKey.Trim();
        _cachedApiKey = clean;

        try
        {
            var setting = _db.SystemSettings.FirstOrDefault(s => s.Key == "gemini_api_key");
            if (setting == null)
            {
                _db.SystemSettings.Add(new SystemSetting
                {
                    Key = "gemini_api_key",
                    Value = clean,
                    UpdatedAt = DateTime.UtcNow
                });
            }
            else
            {
                setting.Value = clean;
                setting.UpdatedAt = DateTime.UtcNow;
            }
            _db.SaveChanges();
            _logger.LogInformation("Gemini API key saved in system_settings database table.");
            return true;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Failed to save Gemini API key in database.");
            return false;
        }
    }

    public bool RemoveApiKey()
    {
        _cachedApiKey = null;
        try
        {
            var setting = _db.SystemSettings.FirstOrDefault(s => s.Key == "gemini_api_key");
            if (setting != null)
            {
                _db.SystemSettings.Remove(setting);
                _db.SaveChanges();
            }
            return true;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Failed to remove Gemini API key from database.");
            return false;
        }
    }

    public async Task<(bool Success, string Message, string? Model)> TestApiKeyAsync(string? testKey = null)
    {
        var keyToUse = !string.IsNullOrWhiteSpace(testKey) ? testKey.Trim() : _cachedApiKey;
        if (string.IsNullOrWhiteSpace(keyToUse))
        {
            return (false, "No API key provided to test. Please enter a Gemini API key.", null);
        }

        try
        {
            var url = $"https://generativelanguage.googleapis.com/v1beta/models/{_model}:generateContent?key={keyToUse}";
            var requestObj = new JsonObject
            {
                ["contents"] = new JsonArray
                {
                    new JsonObject
                    {
                        ["parts"] = new JsonArray
                        {
                            new JsonObject { ["text"] = "Respond with 'OK' if you can read this." }
                        }
                    }
                },
                ["generationConfig"] = new JsonObject
                {
                    ["maxOutputTokens"] = 10
                }
            };

            using var content = new StringContent(requestObj.ToJsonString(), Encoding.UTF8, "application/json");
            using var response = await _httpClient.PostAsync(url, content);

            if (response.IsSuccessStatusCode)
            {
                return (true, $"Connection successful! {_model} is responsive.", _model);
            }

            var err = await response.Content.ReadAsStringAsync();
            try
            {
                using var doc = JsonDocument.Parse(err);
                if (doc.RootElement.TryGetProperty("error", out var errorElem) &&
                    errorElem.TryGetProperty("message", out var msgElem))
                {
                    return (false, $"Gemini API Error ({response.StatusCode}): {msgElem.GetString()}", _model);
                }
            }
            catch { }

            return (false, $"Gemini API returned status {(int)response.StatusCode}: {response.ReasonPhrase}", _model);
        }
        catch (Exception ex)
        {
            return (false, $"Connection error: {ex.Message}", _model);
        }
    }

    public async Task<string> GenerateContentAsync(string prompt, int maxTokens = 4096, string? systemInstruction = null, string? responseMimeType = null)
    {
        if (string.IsNullOrWhiteSpace(_cachedApiKey))
        {
            return "Gemini API key is not configured. Please add your Gemini API key using the 'Gemini API Key' button in the chat header.";
        }

        var url = $"https://generativelanguage.googleapis.com/v1beta/models/{_model}:generateContent?key={_cachedApiKey}";

        var requestObj = new JsonObject
        {
            ["contents"] = new JsonArray
            {
                new JsonObject
                {
                    ["parts"] = new JsonArray
                    {
                        new JsonObject { ["text"] = prompt }
                    }
                }
            }
        };

        var genConfig = new JsonObject
        {
            ["maxOutputTokens"] = maxTokens
        };

        if (!string.IsNullOrWhiteSpace(responseMimeType))
        {
            genConfig["responseMimeType"] = responseMimeType;
        }

        requestObj["generationConfig"] = genConfig;

        if (!string.IsNullOrWhiteSpace(systemInstruction))
        {
            requestObj["systemInstruction"] = new JsonObject
            {
                ["parts"] = new JsonArray
                {
                    new JsonObject { ["text"] = systemInstruction }
                }
            };
        }

        for (int attempt = 0; attempt < 3; attempt++)
        {
            try
            {
                using var content = new StringContent(requestObj.ToJsonString(), Encoding.UTF8, "application/json");
                using var response = await _httpClient.PostAsync(url, content);

                if (response.StatusCode == System.Net.HttpStatusCode.TooManyRequests)
                {
                    return "The AI service has reached its current Gemini API quota. Please wait for the quota to reset or use a Gemini API project with available quota.";
                }

                if (!response.IsSuccessStatusCode)
                {
                    var errorBody = await response.Content.ReadAsStringAsync();
                    _logger.LogWarning("Gemini error (attempt {Attempt}): {Status} - {Body}", attempt, response.StatusCode, errorBody);

                    if ((int)response.StatusCode >= 500 && attempt < 2)
                    {
                        await Task.Delay((int)Math.Pow(2, attempt) * 1000);
                        continue;
                    }

                    return "The AI service could not process the request. Please check the Gemini API configuration.";
                }

                var responseString = await response.Content.ReadAsStringAsync();
                var jsonNode = JsonNode.Parse(responseString);
                var parts = jsonNode?["candidates"]?[0]?["content"]?["parts"]?.AsArray();

                string? text = null;
                if (parts != null)
                {
                    var textParts = new List<string>();
                    foreach (var part in parts)
                    {
                        var t = part?["text"]?.GetValue<string>();
                        if (!string.IsNullOrWhiteSpace(t) && part?["thoughtSignature"] == null)
                        {
                            textParts.Add(t);
                        }
                    }
                    if (textParts.Any())
                    {
                        text = string.Join("", textParts);
                    }
                    else if (parts.Count > 0)
                    {
                        text = parts.LastOrDefault()?["text"]?.GetValue<string>();
                    }
                }

                if (string.IsNullOrWhiteSpace(text))
                {
                    return "I couldn't generate an answer from the available information.";
                }

                return text.Trim();

            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Exception calling Gemini API (attempt {Attempt})", attempt);
                if (attempt >= 2)
                {
                    return "An unexpected error occurred while generating the AI response.";
                }
                await Task.Delay((int)Math.Pow(2, attempt) * 1000);
            }
        }

        return "Sorry, Gemini is currently unavailable.";
    }

    public async IAsyncEnumerable<string> StreamContentAsync(string prompt, int maxTokens = 4096, string? systemInstruction = null, [System.Runtime.CompilerServices.EnumeratorCancellation] CancellationToken cancellationToken = default)
    {
        if (string.IsNullOrWhiteSpace(_cachedApiKey))
        {
            yield return "Gemini API key is not configured. Please add your Gemini API key using the 'Gemini API Key' button in the chat header.";
            yield break;
        }

        var url = $"https://generativelanguage.googleapis.com/v1beta/models/{_model}:streamGenerateContent?alt=sse&key={_cachedApiKey}";

        var requestObj = new JsonObject
        {
            ["contents"] = new JsonArray
            {
                new JsonObject
                {
                    ["parts"] = new JsonArray
                    {
                        new JsonObject { ["text"] = prompt }
                    }
                }
            },
            ["generationConfig"] = new JsonObject
            {
                ["maxOutputTokens"] = maxTokens
            }
        };

        if (!string.IsNullOrWhiteSpace(systemInstruction))
        {
            requestObj["system_instruction"] = new JsonObject
            {
                ["parts"] = new JsonArray
                {
                    new JsonObject { ["text"] = systemInstruction }
                }
            };
        }

        using var content = new StringContent(requestObj.ToJsonString(), Encoding.UTF8, "application/json");
        using var request = new HttpRequestMessage(HttpMethod.Post, url) { Content = content };
        using var response = await _httpClient.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, cancellationToken);

        if (!response.IsSuccessStatusCode)
        {
            var err = await response.Content.ReadAsStringAsync(cancellationToken);
            _logger.LogError("Gemini stream returned error {StatusCode}: {Error}", response.StatusCode, err);
            yield return "Error calling Gemini streaming API.";
            yield break;
        }

        using var stream = await response.Content.ReadAsStreamAsync(cancellationToken);
        using var reader = new StreamReader(stream);

        while (!cancellationToken.IsCancellationRequested)
        {
            var line = await reader.ReadLineAsync(cancellationToken);
            if (line == null) break;
            if (string.IsNullOrWhiteSpace(line)) continue;

            if (line.StartsWith("data: "))
            {
                var json = line.Substring(6).Trim();
                if (json == "[DONE]") break;

                string? chunkText = null;
                try
                {
                    var node = JsonNode.Parse(json);
                    chunkText = node?["candidates"]?[0]?["content"]?["parts"]?[0]?["text"]?.GetValue<string>();
                }
                catch { }

                if (!string.IsNullOrEmpty(chunkText))
                {
                    yield return chunkText;
                }
            }
        }
    }

    public Task<string> AskGeminiAsync(string question, string context)
    {
        var prompt = _promptService.BuildChatPrompt(question, context);
        return GenerateContentAsync(prompt, maxTokens: 4096, systemInstruction: _promptService.ChatSystemInstruction);
    }

    public Task<string> AskSummaryAsync(string context)
    {
        var prompt = _promptService.BuildSummaryPrompt(context);
        return GenerateContentAsync(prompt, maxTokens: 4096);
    }

    public Task<string> AskNotesAsync(string context)
    {
        var prompt = _promptService.BuildNotesPrompt(context);
        return GenerateContentAsync(prompt, maxTokens: 4096);
    }

    public Task<string> AskQuizAsync(string context, string difficulty = "Medium", int questions = 10, List<string>? skills = null)
    {
        var prompt = _promptService.BuildQuizPrompt(context, difficulty, questions, skills);
        return GenerateContentAsync(prompt, maxTokens: 8192, responseMimeType: "application/json");
    }

    public Task<string> AskCourseSkillsAsync(string courseTitle, string context)
    {
        var prompt = _promptService.BuildCourseSkillsPrompt(courseTitle, context);
        return GenerateContentAsync(prompt, maxTokens: 8192, responseMimeType: "application/json");
    }

}
